### PR TITLE
[extractor/facebook] fix extraction

### LINKED ISSUE #30474: cannot download video from facebook url
## Checklist

- [ ] I'm reporting a broken site support issue
- [x] I've verified that I'm running youtube-dl version **2021.12.17**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar bug reports including closed ones
- [x] I've read bugs section in FAQ

## Verbose log

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--verbose', '--username', 'PRIVATE', '--password', 'PRIVATE', '-F', 'https://www.facebook.com/groups/352666925484718/permalink/1112505706167499/']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] youtube-dl version 2021.12.17
[debug] Python version 3.8.10 (CPython) - Linux-5.4.0-91-generic-x86_64-with-glibc2.29
[debug] exe versions: none
[debug] Proxy map: {}
[facebook] Downloading login page
[facebook] Logging in
[facebook] 1112505706167499: Downloading webpage
Traceback (most recent call last):
  File "/home/videodl/vdl/vdl_env/bin/youtube-dl", line 8, in <module>
    sys.exit(main())
  File "/home/videodl/vdl/vdl_env/lib/python3.8/site-packages/youtube_dl/__init__.py", line 475, in main
    _real_main(argv)
  File "/home/videodl/vdl/vdl_env/lib/python3.8/site-packages/youtube_dl/__init__.py", line 465, in _real_main
    retcode = ydl.download(all_urls)
  File "/home/videodl/vdl/vdl_env/lib/python3.8/site-packages/youtube_dl/YoutubeDL.py", line 2068, in download
    res = self.extract_info(
  File "/home/videodl/vdl/vdl_env/lib/python3.8/site-packages/youtube_dl/YoutubeDL.py", line 808, in extract_info
    return self.__extract_info(url, ie, download, extra_info, process)
  File "/home/videodl/vdl/vdl_env/lib/python3.8/site-packages/youtube_dl/YoutubeDL.py", line 815, in wrapper
    return func(self, *args, **kwargs)
  File "/home/videodl/vdl/vdl_env/lib/python3.8/site-packages/youtube_dl/YoutubeDL.py", line 836, in __extract_info
    ie_result = ie.extract(url)
  File "/home/videodl/vdl/vdl_env/lib/python3.8/site-packages/youtube_dl/extractor/common.py", line 534, in extract
    ie_result = self._real_extract(url)
  File "/home/videodl/vdl/vdl_env/lib/python3.8/site-packages/youtube_dl/extractor/facebook.py", line 680, in _real_extract
    return self._extract_from_url(real_url, video_id)
  File "/home/videodl/vdl/vdl_env/lib/python3.8/site-packages/youtube_dl/extractor/facebook.py", line 499, in _extract_from_url
    parse_attachment(attachment)
  File "/home/videodl/vdl/vdl_env/lib/python3.8/site-packages/youtube_dl/extractor/facebook.py", line 480, in parse_attachment
    media = attachment.get(key) or {}
AttributeError: 'NoneType' object has no attribute 'get'

```

## Description

Youtube-dl was working with the same url, username and password. It stopped working recently. video url given in verbose.

### PR DESCRIPTION
**Pull request by kikuyan**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [ ] Covered the code with tests (note that PRs without tests will be REJECTED)
_Maybe later_
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Resolves # 30474 (though already closed)

### CODE REVIEW

File: youtube_dl/extractor/facebook.py [hunk 1]
```diff
@@ -492,7 +492,10 @@ def parse_attachment(attachment, key='media'):
                         lambda x: x['attachments']
                     ], list) or []
                     for attachment in attachments:
-                        attachment = try_get(attachment, lambda x: x['style_type_renderer']['attachment'], dict)
+                        attachment = try_get(attachment, [
```
Comment (dirkf) [inline youtube_dl/extractor/facebook.py:495]: Would it be better style for the getters to be a tuple, immutable, rather than a list?

**Comment by dirkf** [comment]
Aha, so that's the extraction stage that was failing in #30474.

As it's a small change, could it be appropriate to roll it into #29796 as a separate commit?

When I looked at this extractor for #30474, I was concerned that `parse_attachment()`, which expects a dict, is called in various places where that isn't guaranteed. Should the first line of the function be changed:  `media = try_get(attachment, lambda x: x[key], dict) or {}` ?

File: youtube_dl/extractor/facebook.py (hunk 1 above)
Comment (kikuyan) [reply youtube_dl/extractor/facebook.py:495]: This style matches the code around. There also seems to be many use of list in other places like `self._search_regex()` and coding convention examples.
If `for x in ...`, then tuple would be preferred.

**Comment by kikuyan** [comment]
As you said, I'm going to roll this into #29796.
In any case, it is required to test this, and #29796 itself may need additional modifications.
So closing this.

As to `parse_attachment()`: 
I don't know. If the data structure is not as expected, any useful info cannot be gotten, and the result won't be that different anyway, I think.

**Comment by dirkf** [comment]
> As you said, I'm going to roll this into #29796. In any case, it is required to test this, and #29796 itself may need additional modifications. So closing this.
> 
> As to `parse_attachment()`: I don't know. If the data structure is not as expected, any useful info cannot be gotten, and the result won't be that different anyway, I think.

My feeling was that it's better to crash

**Comment by kikuyan** [comment]
I feel it's excessive, since the code was written by analyzing data structures.

"No video formats found", or some other unexpected error will be raised, "please report this issue on h<span></span>ttps://yt-dl.org/bug" will be shown next.
Users will make issue reports anyway, so no difference than just crash I think, except a bit ugly.

**Pull request closed without merging**